eth/downloader/skeleton.go

```diff
@@ -0,0 +1,1013 @@
+			tail := s.progress.Subchains[0].Tail
+			want := s.progress.Subchains[0].Next
+			have := s.scratchSpace[0].Hash()
+
+			log.Warn("Invalid skeleton headers", "peer", s.scratchOwners[0], "number", tail-1, "want", want, "have", have)
+
+			// The peer delivered junk, or at least not the subchain we are
+			// syncing to. Free up the scratch space and assignment, reassign
+			// and drop the original peer.
+			for i := 0; i < requestHeaders; i++ {
+				s.scratchSpace[i] = nil
+			}
+			s.drop(s.scratchOwners[0])
+			s.scratchOwners[0] = ""
+			break
+		}
+		// Scratch delivery matches required subchain, deliver the batch of
+		// headers and push the subchain forward
+		var consumed int
+		for _, header := range s.scratchSpace[:requestHeaders] {
+			if header != nil { // nil when the genesis is reached
+				consumed++
+
+				rawdb.WriteSkeletonHeader(batch, header)
+				s.pulled++
+
+				s.progress.Subchains[0].Tail--
+				s.progress.Subchains[0].Next = header.ParentHash
+			}
+		}
+		// Batch of headers consumed, shift the download window forward
+		head := s.progress.Subchains[0].Head
+		tail := s.progress.Subchains[0].Tail
+		next := s.progress.Subchains[0].Next
+
+		log.Trace("Primary subchain extended", "head", head, "tail", tail, "next", next)
+
+		copy(s.scratchSpace, s.scratchSpace[requestHeaders:])
+		for i := 0; i < requestHeaders; i++ {
+			s.scratchSpace[scratchHeaders-i-1] = nil
+		}
+		copy(s.scratchOwners, s.scratchOwners[1:])
+		s.scratchOwners[scratchHeaders/requestHeaders-1] = ""
+
+		s.scratchHead -= uint64(consumed)
+
+		// If the subchain extended into the next subchain, we need to handle
+		// the overlap. Since there could be many overlaps (come on), do this
+		// in a loop.
+		for len(s.progress.Subchains) > 1 && s.progress.Subchains[1].Head >= s.progress.Subchains[0].Tail {
+			// Extract some stats from the second subchain
+			head := s.progress.Subchains[1].Head
+			tail := s.progress.Subchains[1].Tail
+			next := s.progress.Subchains[1].Next
+
+			// Since we just overwrote part of the next subchain, we need to trim
+			// its head independent of matching or mismatching content
+			if s.progress.Subchains[1].Tail >= s.progress.Subchains[0].Tail {
+				// Fully overwritten, get rid of the subchain as a whole
+				log.Debug("Previous subchain fully overwritten", "head", head, "tail", tail, "next", next)
+				s.progress.Subchains = append(s.progress.Subchains[:1], s.progress.Subchains[2:]...)
+				continue
+			} else {
+				// Partially overwritten, trim the head to the overwritten size
+				log.Debug("Previous subchain partially overwritten", "head", head, "tail", tail, "next", next)
+				s.progress.Subchains[1].Head = s.progress.Subchains[0].Tail - 1
+			}
+			// If the old subchain is an extension of the new one, merge the two
+			// and let the skeleton syncer restart (to clean internal state)
+			if rawdb.ReadSkeletonHeader(s.db, s.progress.Subchains[1].Head).Hash() == s.progress.Subchains[0].Next {
```

Comment:
I'm not fully convinced that this is correct. At this point, the batch hasn't been written, but the read reads from the underlying db. We keep pushing things into the batch (not writing to db). I have a hard time figuring out if this will work when a lot of merged deliveries are made

Reply:
We're reading a header belonging to a previously existing and aborted subchain. All data should be present for that old subchain. I.e. currently we *never* delete a beacon header, only ever overwrite it. So if an interval is covered, *something* should exist there,